File: src/bc-process-pool.ts
/**
 * BC Process Pool Manager
 * Manages a pool of BC processes for concurrent request handling
 */

import { BCProcess } from './bc-process.js';
import { ProcessPoolConfig, BCCalculatorError, BCErrorCode } from './types.js';

export class BCProcessPool {
  private processes: BCProcess[] = [];
  private availableProcesses: BCProcess[] = [];
  private config: ProcessPoolConfig;
  private isInitialized: boolean = false;

  constructor(config: ProcessPoolConfig) {
    this.config = config;
  }

  /**
   * Initialize the process pool by spawning configured number of BC processes
   */
  async initialize(): Promise<void> {
    if (this.isInitialized) {
      console.error('BC process pool already initialized');
      return;
    }

    console.error(`Initializing BC process pool with ${this.config.poolSize} processes...`);

    try {
      // Create pool of processes
      const promises: Promise<BCProcess>[] = [];
      for (let i = 0; i < this.config.poolSize; i++) {
        promises.push(this.createProcess(i));
      }
      
      // Wait for all processes to start
      await Promise.all(promises);
      
      this.isInitialized = true;
      console.error(`BC process pool initialized successfully with ${this.processes.length} processes`);
    } catch (error) {
      console.error(`Failed to initialize BC process pool: ${error}`);
      // Clean up any processes that were created
      await this.shutdown();
      throw new BCCalculatorError(
        `Failed to initialize BC process pool: ${error instanceof Error ? error.message : String(error)}`,
        BCErrorCode.BC_SPAWN_ERROR,
        error
      );
    }
  }

  /**
   * Create and initialize a new BC process
   * @param index Process index for logging
   */
  private async createProcess(index?: number): Promise<BCProcess> {
    const logPrefix = index !== undefined ? `[Process ${index}]` : '[Process]';
    
    try {
      const process = new BCProcess({
        precision: this.config.defaultPrecision,
        timeout: this.config.defaultTimeout,
        useMathLibrary: true
      });
      
      await process.start();
      
      this.processes.push(process);
      this.availableProcesses.push(process);
      
      if (index !== undefined) {
        console.error(`${logPrefix} Started successfully (PID: ${process.getPid()})`);
      }
      
      return process;
    } catch (error) {
      console.error(`${logPrefix} Failed to start: ${error}`);
      throw error;
    }
  }

  /**
   * Acquire a process from the pool
   * Waits if no processes are currently available
   */
  async acquireProcess(): Promise<BCProcess> {
    if (!this.isInitialized) {
      throw new BCCalculatorError(
        'BC process pool not initialized',
        BCErrorCode.BC_NOT_READY
      );
    }

    // Wait for an available process
    const maxWaitTime = 30000; // 30 seconds
    const checkInterval = 100; // Check every 100ms
    let waitedTime = 0;

    while (this.availableProcesses.length === 0 && waitedTime < maxWaitTime) {
      await new Promise(resolve => setTimeout(resolve, checkInterval));
      waitedTime += checkInterval;
    }

    if (this.availableProcesses.length === 0) {
      throw new BCCalculatorError(
        'No BC processes available after waiting',
        BCErrorCode.BC_NOT_READY
      );
    }

    const process = this.availableProcesses.shift()!;
    
    // Verify process is still healthy
    if (!process.isAvailable()) {
      console.error(`Process ${process.getPid()} is not available, attempting recovery`);
      // Process died, create replacement
      process.kill();
      const newProcess = await this.createProcess();
      return newProcess; // New process is already available
    }

    return process;
  }

  /**
   * Release a process back to the pool
   * If the process is unhealthy, it will be replaced
   */
  releaseProcess(process: BCProcess): void {
    if (!this.processes.includes(process)) {
      console.error('Attempted to release process not in pool');
      return;
    }

    if (process.isAvailable()) {
      // Process is healthy, return to available pool
      if (!this.availableProcesses.includes(process)) {
        this.availableProcesses.push(process);
      }
    } else {
      // Process is unhealthy, replace it
      console.error(`Process ${process.getPid()} is unhealthy, replacing...`);
      
      // Remove from both lists
      this.processes = this.processes.filter(p => p !== process);
      this.availableProcesses = this.availableProcesses.filter(p => p !== process);
      
      // Kill the bad process
      process.kill();
      
      // Create replacement (non-blocking)
      this.createProcess().catch(error => {
        console.error(`Failed to create replacement process: ${error}`);
      });
    }
  }

  /**
   * Get the number of available processes
   */
  getAvailableCount(): number {
    return this.availableProcesses.length;
  }

  /**
   * Get the total number of processes in the pool
   */
  getTotalCount(): number {
    return this.processes.length;
  }

  /**
   * Get pool status information
   */
  getStatus(): {
    initialized: boolean;
    totalProcesses: number;
    availableProcesses: number;
    busyProcesses: number;
  } {
    return {
      initialized: this.isInitialized,
      totalProcesses: this.processes.length,
      availableProcesses: this.availableProcesses.length,
      busyProcesses: this.processes.length - this.availableProcesses.length
    };
  }

  /**
   * Shutdown the process pool
   * Kills all BC processes
   */
  async shutdown(): Promise<void> {
    console.error('Shutting down BC process pool...');
    
    for (const process of this.processes) {
  process.kill();
    }
    
    this.processes = [];
    this.availableProcesses = [];
    this.isInitialized = false;
    
    console.error('BC process pool shutdown complete');
  }
}

File: src/python/bc_calculator_mcp/bc_process_pool.py
"""
BC Process Pool Manager
Manages a pool of BC processes for concurrent request handling
"""

import asyncio
import sys
from typing import List

from .bc_process import BCProcess
from .types import BCCalculatorError, BCErrorCode, BCProcessOptions, ProcessPoolConfig


class BCProcessPool:
    """Manager for a pool of BC processes with concurrent access control"""
    
    def __init__(self, config: ProcessPoolConfig) -> None:
        self._config = config
        self._processes: List[BCProcess] = []
        self._available: asyncio.Queue[BCProcess] = asyncio.Queue()
        self._semaphore = asyncio.Semaphore(config.pool_size)
        self._is_initialized: bool = False
    
    async def initialize(self) -> None:
        """Initialize the process pool by spawning configured number of BC processes"""
        if self._is_initialized:
            print("BC process pool already initialized", file=sys.stderr)
            return
        
        print(f"Initializing BC process pool with {self._config.pool_size} processes...", file=sys.stderr)
        
        try:
            # Create pool of processes
            for i in range(self._config.pool_size):
                process = await self._create_process(i)
                self._processes.append(process)
                await self._available.put(process)
            
            self._is_initialized = True
            print(f"BC process pool initialized successfully with {len(self._processes)} processes", file=sys.stderr)
            
        except Exception as error:
            print(f"Failed to initialize BC process pool: {error}", file=sys.stderr)
            # Clean up any processes that were created
            await self.shutdown()
            raise BCCalculatorError(
                f"Failed to initialize BC process pool: {str(error)}",
                BCErrorCode.BC_SPAWN_ERROR,
                error
            )
    
    async def _create_process(self, index: int = -1) -> BCProcess:
        """
        Create and initialize a new BC process
        
        Args:
            index: Process index for logging
            
        Returns:
            Initialized BCProcess
            
        Raises:
            BCCalculatorError: If process creation fails
        """
        log_prefix = f"[Process {index}]" if index >= 0 else "[Process]"
        
        try:
            process = BCProcess(
                BCProcessOptions(
                    precision=self._config.default_precision,
                    timeout=self._config.default_timeout,
                    use_math_library=True
                )
            )
            
            await process.start()
            
            if index >= 0:
                print(f"{log_prefix} Started successfully (PID: {process.get_pid()})", file=sys.stderr)
            
            return process
            
        except Exception as error:
            print(f"{log_prefix} Failed to start: {error}", file=sys.stderr)
            raise
    
    async def acquire_process(self) -> BCProcess:
        """
        Acquire a process from the pool
        Waits if no processes are currently available
        
        Returns:
            Available BCProcess
            
        Raises:
            BCCalculatorError: If pool not initialized or timeout waiting
        """
        if not self._is_initialized:
            raise BCCalculatorError(
                "BC process pool not initialized",
                BCErrorCode.BC_NOT_READY
            )
        
        # Wait for an available process (with timeout)
        try:
            process = await asyncio.wait_for(
                self._available.get(),
                timeout=30.0  # 30 second timeout
            )
        except asyncio.TimeoutError:
            raise BCCalculatorError(
                "No BC processes available after waiting 30 seconds",
                BCErrorCode.BC_NOT_READY
            )
        
        # Verify process is still healthy
        if not process.is_available():
            print(f"Process {process.get_pid()} is not available, attempting recovery", file=sys.stderr)
            # Process died, create replacement
            process.kill()
            
            # Remove from processes list
            if process in self._processes:
                self._processes.remove(process)
            
            # Create new process
            new_process = await self._create_process()
            self._processes.append(new_process)
            return new_process
        
        return process
    
    def release_process(self, process: BCProcess) -> None:
        """
        Release a process back to the pool
        If the process is unhealthy, it will be replaced
        
        Args:
            process: The process to release
        """
        if process not in self._processes:
            print("Attempted to release process not in pool", file=sys.stderr)
            return
        
        if process.is_available():
            # Process is healthy, return to available pool
            try:
                self._available.put_nowait(process)
            except asyncio.QueueFull:
                # This shouldn't happen, but handle it gracefully
                print("Warning: Available queue is full", file=sys.stderr)
        else:
            # Process is unhealthy, replace it
            print(f"Process {process.get_pid()} is unhealthy, replacing...", file=sys.stderr)
            
            # Remove from processes list
            if process in self._processes:
                self._processes.remove(process)
            
            # Kill the bad process
            process.kill()
            
            # Create replacement (non-blocking)
            asyncio.create_task(self._replace_process())
    
    async def _replace_process(self) -> None:
        """Create a replacement process asynchronously"""
        try:
            new_process = await self._create_process()
            self._processes.append(new_process)
            await self._available.put(new_process)
        except Exception as error:
            print(f"Failed to create replacement process: {error}", file=sys.stderr)
    
    def get_available_count(self) -> int:
        """Get the number of available processes"""
        return self._available.qsize()
    
    def get_total_count(self) -> int:
        """Get the total number of processes in the pool"""
        return len(self._processes)
    
    def get_status(self) -> dict:
        """
        Get pool status information
        
        Returns:
            Dictionary with pool status
        """
        return {
            "initialized": self._is_initialized,
            "totalProcesses": len(self._processes),
            "availableProcesses": self._available.qsize(),
            "busyProcesses": len(self._processes) - self._available.qsize()
        }
    
    async def shutdown(self) -> None:
        """
        Shutdown the process pool
        Kills all BC processes
        """
        print("Shutting down BC process pool...", file=sys.stderr)
        
        for process in self._processes:
            process.kill()
        
        self._processes = []
        # Clear the queue
        while not self._available.empty():
            try:
                self._available.get_nowait()
            except asyncio.QueueEmpty:
                break
        
        self._is_initialized = False
        
        print("BC process pool shutdown complete", file=sys.stderr)